dd-sdk-android/src/main/kotlin/com/datadog/android/core/internal/persistence/file/batch/BatchFileDataReader.kt

```diff
@@ -0,0 +1,115 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2016-Present Datadog, Inc.
+ */
+
+package com.datadog.android.core.internal.persistence.file.batch
+
+import com.datadog.android.core.internal.data.file.Batch
+import com.datadog.android.core.internal.domain.PayloadDecoration
+import com.datadog.android.core.internal.persistence.DataReader
+import com.datadog.android.core.internal.persistence.file.FileHandler
+import com.datadog.android.core.internal.persistence.file.FileOrchestrator
+import com.datadog.android.log.Logger
+import java.io.File
+import java.util.Locale
+
+/**
+ * A [DataReader] reading [Batch] data from files.
+ */
+internal class BatchFileDataReader(
+    internal val fileOrchestrator: FileOrchestrator,
+    internal val decoration: PayloadDecoration,
+    internal val handler: FileHandler,
+    internal val internalLogger: Logger
+) : DataReader {
+
+    private val lockedFiles: MutableList<File> = mutableListOf()
+
+    // region DataReader
+
+    override fun lockAndReadNext(): Batch? {
+        val file = getAndLockReadableFile() ?: return null
+        val data = handler.readData(file, decoration.prefixBytes, decoration.suffixBytes)
+
+        return Batch(file.name, data)
+    }
+
+    override fun release(data: Batch) {
+        releaseFile(data.id, delete = false)
+    }
+
+    override fun drop(data: Batch) {
+        releaseFile(data.id, delete = true)
+    }
+
+    override fun dropAll() {
+        synchronized(lockedFiles) {
+            lockedFiles.toTypedArray().forEach {
+                releaseFile(it, delete = true)
+            }
+        }
+
+        fileOrchestrator.getAllFiles().forEach {
+            deleteFile(it)
+        }
+    }
+
+    // endregion
+
+    // region Internal
+
+    private fun getAndLockReadableFile(): File? {
+        synchronized(lockedFiles) {
+            val readableFile = fileOrchestrator.getReadableFile(lockedFiles.toSet())
+            if (readableFile != null) {
+                lockedFiles.add(readableFile)
+            }
+            return readableFile
+        }
+    }
+
+    private fun releaseFile(
+        fileName: String,
+        delete: Boolean
+    ) {
+        val file = synchronized(lockedFiles) {
+            lockedFiles.firstOrNull { it.name == fileName }
+        }
+        if (file != null) {
+            releaseFile(file, delete)
+        } else {
+            internalLogger.w(
```

Comment:
In theory it won't since the releaseFile will always be called from the upload thread, which is unique. And those log will actually tell us if this is more common than we thought.